lightning/src/ln/interactivetxs.rs

```diff
@@ -430,6 +418,119 @@ impl InteractiveTxSigningSession {
 			output: outputs.iter().cloned().map(|output| output.into_tx_out()).collect(),
 		}
 	}
+
+	pub fn update_from_funding_tx_after_read(&mut self, funding_tx: &Transaction) {
+		self.unsigned_tx.inputs.iter_mut().zip(funding_tx.input.iter().cloned()).for_each(
+			|(placeholder, real)| {
+				*placeholder.txin_mut() = real;
+			},
+		);
+		self.unsigned_tx.outputs.iter_mut().zip(funding_tx.output.iter().cloned()).for_each(
+			|(placeholder, real)| {
+				(*placeholder).output = OutputOwned::Single(real);
+			},
+		);
+	}
+}
+
+impl Writeable for InteractiveTxSigningSession {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
+		let input_serial_ids =
+			self.unsigned_tx.inputs.iter().map(|input| input.serial_id()).collect::<Vec<_>>();
+		let output_serial_ids =
+			self.unsigned_tx.outputs.iter().map(|output| output.serial_id).collect::<Vec<_>>();
+
+		write_tlv_fields!(writer, {
+			(0, self.unsigned_tx.holder_is_initiator, required),
+			(2, self.unsigned_tx.holder_sends_tx_signatures_first, required),
+			(4, self.has_received_commitment_signed, required),
+			(6, input_serial_ids, required),
+			(8, output_serial_ids, required),
+			(10, self.holder_tx_signatures, required),
+		});
+		Ok(())
+	}
+}
+
+impl Readable for InteractiveTxSigningSession {
+	fn read<R: io::Read>(reader: &mut R) -> Result<Self, DecodeError> {
+		let mut holder_is_initiator = false;
+		let mut holder_sends_tx_signatures_first = false;
+		let mut has_received_commitment_signed = false;
+		let mut input_serial_ids: Vec<SerialId> = vec![];
+		let mut output_serial_ids: Vec<SerialId> = vec![];
+		let mut holder_tx_signatures: Option<TxSignatures> = None;
+
+		read_tlv_fields!(reader, {
+			(0, holder_is_initiator, required),
+			(2, holder_sends_tx_signatures_first, required),
+			(4, has_received_commitment_signed, required),
+			(6, input_serial_ids, required),
+			(8, output_serial_ids, required),
+			(10, holder_tx_signatures, required),
+		});
+
+		// We use an empty dummy `TxOut` for all `TxOut`s as they will be populated with their real
+		// values from `funding_transaction` after reading.
```

Comment:
Won't this not be the case if `funding_transaction` is `None`?

Reply:
I've walked back the "minimal" approach to persisting the signing session. We're losing information we'll actually need like adding roles and shared output amounts. I've implemented persistence for `ConstructedTransaction` to keep that information.